invenio_vocabularies/services/facets/facets.py

```diff
@@ -0,0 +1,129 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2021 CERN.
+#
+# Invenio-Vocabularies is free software; you can redistribute it and/or
+# modify it under the terms of the MIT License; see LICENSE file for more
+# details.
+
+"""Vocabulary facets."""
+
+from bisect import bisect, bisect_left
+
+from elasticsearch_dsl import AttrDict
+from invenio_records_resources.services.records.facets import TermsFacet
+
+
+class LabellingBucket(AttrDict):
+    """Bucket class to allow label extration."""
+
+    def __init__(self, key, doc_count):
+        """Constructor."""
+        super().__init__({"key": key, "doc_count": doc_count})
+
+
+class NestedVocabularyTermsFacet(TermsFacet):
+    """A term facet for vocabularies with nested/hierarchical labelling."""
+
+    def __init__(self, splitchar="-", **kwargs):
+        """Initialize the nested vocabulary terms facet."""
+        self._splitchar = splitchar
+        super().__init__(**kwargs)
+
+    def _build_nested_buckets(self, buckets):
+        """Build nested buckets from a vocabulary facet."""
+        seen = set()  # cheaper insert and in operations
+        buckets_idx = []  # avoid implementing __lt__ for bisect on buckets
+        buckets_out = []  # sorted, non duplicates buckets
+        for bucket in buckets:
+            key = self.get_value(bucket)
+            metric = self.get_metric(bucket)
+            nested = key.split(self._splitchar)
+
+            root = ""
+            for facet in nested[:-1]:
+                root = facet if not root else root + "-" + facet
+                if root not in seen:
+                    seen.add(root)
+                    idx = bisect(buckets_idx, root)
+                    buckets_out.insert(idx, LabellingBucket(root, metric))
+                    buckets_idx.insert(idx, root)  # add root to idx
+                else:  # update aggregated count
+                    idx = bisect_left(buckets_idx, root)
+                    buckets_out[idx].doc_count += metric
+
+            if key not in seen:  # Add the last bucket
+                seen.add(key)
+                idx = bisect(buckets_idx, key)
+                buckets_out.insert(idx, bucket)
+                buckets_idx.insert(idx, key)  # add full key to idx
+
+        return buckets_out
```

Comment:
I am sure there are a lot of smart things happening here, but it is kind of hard to figure out what this list will contain at the end, in particular with nested buckets.

For these given buckets:
* image
* image-png
* image-jpg

Is it returning `[image, image-png, image-jpg]`?

Other question: are the bucket returned by ES alphabetically sorted? Or sort by metric descending? Do we have to re-sort them, for example to have instead: `[image, image-jpg, image-png]`
This is what I would try to add to the method docstring to understand better what is happening here.

Reply:
The comment is correct. Imagine this two cases:

```
buckets.data = [ image-png, image-jpg, publication]
buckets_out = [image, image-png, image-jpg, publication]
```

and

```
buckets.data = [image, image-png, image-jpg, publication]
buckets_out = [image, image-png, image-jpg, publication]
```

As for the order, they do not... that's why I am using the bisect library to be able to have a sorted `bucket_out` and need to maintain a sorted `bucket_idx` with just the key (not the buckets). Otherwise, I would need to re-implement the `__lt__` function so bisect can properly find the insertion point.